src/interfaces/vault.ts

```diff
@@ -330,11 +330,16 @@ export class VaultInterface<T extends ChainId> extends ServiceInterface<T> {
       if (token === EthAddress) {
         throw new SdkError("deposit:v2:eth not implemented");
       } else {
+        const shouldUsePartner = this.yearn.services.partner && this.yearn.services.partner.isAllowed(vault);
         const vaultContract = new Contract(vault, VaultAbi, signer);
+
         const makeTransaction = async (overrides: CallOverrides) => {
-          const tx = await vaultContract.populateTransaction.deposit(amount, overrides);
+          const tx = shouldUsePartner
+            ? await this.yearn.services.partner!.populateDepositTransaction(vault, amount, overrides)
```

Comment:
The bang here feels dangerous, we should avoid that at all costs.

Could we refactor so that when TypeScript gets here it knows that we have a `this.yearn.services.partner`? We are checking that in line 333, does it still think that it might be `undefined`?

Reply:
Yeah the issue here is that typescript does not correctly detect that this.yearn.services.partner will exist at this point.

If we change shouldUsePartner to directly checking `this.yearn.services.partner` it would correctly say "ok there is no way this is going to be undefined". I get the worries about the bangs, but if used accordingly and causiosly they are fine

Reply:
I think we need to figure why TypeScript is not able to infer that `this.yearn.services.partner` is defined and then remove the bang. I really think there should almost never be a need for it

Reply:
Seems like its a problem with the version of typescript https://stackoverflow.com/questions/65050560/typescript-type-inference-in-an-if-statement-doesnt-work-as-a-variable

We use version 3.9 while that change was introduced in version 4.4. I could add a `shouldUsePartner && this.yearn.services.partner` but it seems to be just a workaround for how weirdly typescript works and I don't think we should abide to that

Reply:
Nice find! All good then, let's go with what we have.

btw, I'm unable to resolve conversations 😢 @jstashh please give me the permission